### ISSUE TITLE
Indicators for current/previous tab in :buffer completion

### ISSUE THREAD

**Issue by haasn** [The-Compiler/qutebrowser#1778]
In pentadactyl I've grown rather attached to the list of open buffers, which I can open by pressing B. Especially since I like to have very many open tabs, and I often check their titles (even in situations where I don't want to navigate them), I would appreciate a command similar to this; plus the ability to turn the tab bar itself off.

Here's a screenshot of my desired state:
![buffers](https://0x0.st/BZm.png)

The current :buffer command is already very similar to this in practice, so I imagine appeasing my weird taste shouldn't be too difficult :p

Mostly I'm missing the favicon, I think - and the ability to exit the buffer list using `<CR>` - but ideally I'd want the buffer list to automatically update itself if a buffer changes its title while the buffer list is open.

**Comment by The-Compiler** [comment]
> plus the ability to turn the tab bar itself off

`:set tabs show false` :wink:

> Mostly I'm missing the favicon, I think

I opened #1784 for that.

> and the ability to exit the buffer list using <CR>

You can do that (but will get an error), or you can just press escape.

> but ideally I'd want the buffer list to automatically update itself if a buffer changes its title while the buffer list is open.

I'd expect it to do that, and if not, it should be relatively straight-forward to add. Do you have an example-page by any chance?

**Comment by haasn** [comment]
You mean an example page that updates its own title/favicon? I'm afraid not - I mainly use it for stuff like message boards and social networking sites with their own auto-updating/notification systems which change the title and favicon (e.g. adding an exclamation mark) to notify me about updates.

One of the highest-volume websites I know that has this behavior (if you turn on automatic updates on a thread) is 4chan.org (content warning!), so you could probably use that for testing the behavior reasonably quickly, if nothing else.

**Comment by haasn** [comment]
Another minor thing I noticed which the pentadactyl buffer list provides: an indication of what the current tab is (%) along with what the previously selected tab was (#).

This is actually the most important feature that I only just realized is missing from the current buffer list.

**Comment by The-Compiler** [comment]
@blyxxyz fixed updating of the title in #1798, and I added a test page in 4505d74ae32c8727b02e871dd856ec320a1f29a2.

If I'm not mistaken, the only open item here are the current/previous tab indicators in the completion - correct?

**Comment by haasn** [comment]
I think so. The presentation of the qutebrowser buffer list could still use some work (IMO), but all of that should mostly already be covered by other issues:

* favicons (covered in #1784)
* color distinction (covered in #1794)
* maybe a different order? (I think pentadactyl's `number, icon, title, url` order is the most natural, but I'm obviously biased since I'm pretty much just trying to completely recreate pentadactyl in case it wasn't obvious :p)

**Comment by The-Compiler** [comment]
Renaming this then :wink:

> maybe a different order? (I think pentadactyl's `number, icon, title, url` order is the most natural, but I'm obviously biased since I'm pretty much just trying to completely recreate pentadactyl in case it wasn't obvious :p)

I tend to disagree, especially because the format is "url, title" in other completions too.

**Comment by haasn** [comment]
@The-Compiler Well, I think title, url makes more sense in all contexts - not just the buffer list. To me the title of a web page tends to say more about the semantic meaning of it than the URL, but I imagine this is rather up for preference.

Maybe it would be a good idea to approach this by giving each unique “type” of column a configurable “priority” (simple integer), alongside the idea I had for giving them a configurable color. (As an added benefit, I could imagine for example setting the priority to 0 to disable a column type I am less interested in)

In summary, the completion list in general could be reworked to where every type of completion only provides an effectively unordered list of typed columns it can provide for values - with the order and presentation being determined by user preference (priority, color, more?).

**Comment by haasn** [comment]
Ah, another pretty big difference between qute's buffer list and pentadactyl's: pentadactyl's only shows buffers from the current window.

I'm not sure what good showing me buffers from another window is supposed to do in `:buffer`, since it's not like selecting them does anything. I think that maybe this should be a flag to `:buffer`, and off by default.

**Comment by The-Compiler** [comment]
Split off to #1808 :wink:

**Comment by haasn** [comment]
De-prioritizing this for myself since I've grown to like the tab bar + many windows concept more than the old “one window with a bazillion buffers” that I employed using pentadactyl.

This paradigm shift is partly due to the tab bar in qutebrowser being significantly better than in firefox (I always had the tab bar hidden in firefox with basically no choice over the matter), and partly due to the fact that qutebrowser is far the only browser I've seen in which creating new windows is actually a free action.

I still think the “set column type priority” thing is useful for :history and (eventually) :undo, though, so I wouldn't count that out just yet - but it's probably best to merge that in with #1794 and close this issue.

**Comment by The-Compiler** [comment]
I still think the indicators would be a nice visual thing, but I'm going to merge this into #1784 which is quite related.
